In webgl drawer, fall back to canvas drawer for tiled images with tainted data

pearcetm opened the pull request:
Displaying images from cross-origin sources that don't have CORS enabled because of viewer config (or server settings, I guess) is a problem for the new `WebGLDrawer` - tainted data cannot be uploaded to `WebGL`. See https://github.com/openseadragon/openseadragon/issues/2470 for example.

This PR adds the ability for the `WebGLDrawer` to detect and lazily instantiate a hidden `CanvasDrawer` to use for `TiledImage`s that have tainted data. The `WebGLDrawer` will still draw any images it can, and both tainted and non-tainted images can be rendered in the same viewer. When this happens, a warning message is logged to the console.

To test, visit the `drawercomparison` demo. In the source, comment out `crossOriginPolicy` for the viewer with `drawer: 'webgl'` here:
https://github.com/openseadragon/openseadragon/blob/f7d86ac244826416d3090fbef86968f17fc40fe8/test/demo/drawercomparison.js#L52
Toggle on the Duomo image (which is cross-origin), and a `CanvasDrawer` will be created and used for rendering specifically this image.

pearcetm commented:
I've gone back and forth on adding new APIs around this functionality.

One idea would be to add an option during viewer creation that would enable or disable this fallback behavior. If we do this, I would vote for enabling it by default so that the majority of existing apps continue to "just work" even if they are using cross-origin images without enabling CORS.

A second idea would be to raise a new event when a `TiledImage` is marked as `tainted`. This would give developers a hook to use to handle the situation in some way or other. For example, maybe they would want to display a warning that some particular `WebGL` functionality can't be used, and disable a filtering UI or something like that.

Aiosa commented:
I am for the event. Also, regarding options I added base drawer option functionality to the cache PR (which probably would be better in their own smaller request :D )

Aiosa commented:
Basically any drawer should be able to perform test of functionality and raise event on failure... Base class stuff.

pearcetm commented on src/webgldrawer.js:
```diff
@@ -237,169 +257,188 @@
-                // First rendering pass: compose tiles that make up this tiledImage
-                gl.useProgram(this._firstPass.shaderProgram);
+                    // next, use the backup canvas drawer to draw this tainted image
+                    const canvasDrawer = this._getBackupCanvasDrawer();
```
For applications that have potentially mixed tile sources, I want to be able to use the "best" drawer (i.e. the one specified when creating the viewer") for any `TiledImage`s that it the drawer is able to draw, and only fall back when necessary, rather than downgrading all images by replacing the entire drawer.

If a developer wants to, they could listen for the to-be-implemented `drawer-failure` event (or whatever we name it) and replace the whole drawer themselves.

Aiosa replied:
My point was:
- it is currently not easy to replace drawer, you would have to probably recreate the whole OSD instance so having implemented this fallback would help users too, supposing OSD re-instantion is not enough
- in most cases, if one of your data sources is tained, all of them are
- vanila WebGL drawing is less performant than canvas for obivous reasons, so that when you use webgl renderer, you
   - either want to get rid of seams -> then you get the seams with faulty data anyway
   - want to employ custom GPU processing features -> then the fallback per item image does not work anyway

So to me it seems like unnecessary feature, and supporting general drawer fallback replacement on the global level is a) cleaner and b) more 'stable' solution...

pearcetm replied:
> * it is currently not easy to replace drawer,

It actually shouldn't be too hard to swap out the drawer. Maybe we should add an API to do so.

> * in most cases, if one of your data sources is tained, all of them are

There are plenty of use-cases where an OSD viewer can be used to view images from across the web, which are not all under the control of the front-end developer. I think it would be a mistake to totally change the viewer away from webgl as soon as a single image is opened that doesn't have CORS configured.

> So to me it seems like unnecessary feature, and supporting general drawer fallback replacement on the global level is a) cleaner and b) more 'stable' solution...

I disagree, for the above reasons. The fact that the `CanvasDrawer` itself is used as the fallback (rather than copy/pasting code from that drawer) makes this approach pretty robust.

pearcetm commented on src/webgldrawer.js:
```diff
@@ -802,7 +841,19 @@
+            // if the tile doesn't provide a canvas, or is tainted by cross-origin
+            // data, marked the TiledImage as tainted so the canvas drawer can be
+            // used instead, and return immediately - data cannot be uploaded to webgl
+            if(!canvas || $.isCanvasTainted(canvas)){
```
This is just where `isTainted` is set on the `TiledImage`. This is the first place the drawer accesses data from a loaded tile, so I think this is the logical place to test for tainted status. Once a single tile is found to be tainted, the whole `TiledImage` is marked as tainted, and thus it will always be skipped by `WebGLDrawer` and drawn instead by `CanvasDrawer`.

An alternative would be to test each tile that makes up a frame, every frame... but that seemed too expensive.

pearcetm replied:
I feel like the case where only *some* tiles for a given `TiledImage` are tainted is a pretty unlikely edge case. However, I added a guard against loading tiles from tainted `TiledImage`s just in case. If those tiles are shared by multiple `TiledImage`s, and some of them aren't tainted for some reason (super-duper edge case) they will be loaded as soon as necessary for drawing the non-tainted image.

Unloading existing non-tainted tiles should be left to the time they are unloaded from memory, because of the shared-tile edge case.

pearcetm commented on src/webgldrawer.js:
```diff
@@ -206,6 +207,25 @@
+                this._backupCanvasDrawer = new $.CanvasDrawer({
+                    viewer: this.viewer,
+                    viewport: this.viewport,
+                    element: this.container,
```
`debugGridColor` is the only "missing" option as far as I can tell, and it is optional. I'm not sure whether passing through the `WebGLDrawer`'s value for this option would make sense or not.

My goal is to provide a reasonable fallback drawing implementation, but if a default debug color is used instead of the dev-specified one I hope they will just see the warning message in the console and fix their CORS settings first.

Aiosa replied:
I was having additional parameters in the cache refactoring shared by all drawers, that's why I mention it here.

iangilman replied:
@Aiosa What would you propose here to make sure the additional parameters are passed through?

Aiosa replied:
to provide the target options from the `drawerOptions` map... with new cache system a drawer can e.g. specify whether to store its data as internal cache, or re-use the 'main' tile cache, which can be toggled by users exactly through these options, and all dravers should get them

iangilman replied:
Cool. What do you think @pearcetm?

pearcetm replied:
I think it's a good idea. I'm planning to consolidate the drawer creation code into an internal private API on the viewer so that the same initialization and options are always used. I'm also planning to add a `drawer-error` event (I'm open to suggestions on the naming of this event) that can be fired when something goes wrong with drawing a particular TiledImage. After that I'll fix up whatever merge conflicts there are. It may take me a week or more to get to these things.

iangilman replied:
@pearcetm Sounds good!

pearcetm commented:
I added a new API `viewer.setDrawer` that allows the drawer implementation to be easily changed on the fly. For example, in any of the demos, you can call `viewer.setDrawer('canvas')` from the console (assuming window.viewer exists) to switch over to `canvas` from the default `webgl`. Anything that can be passed in during viewer construction can be used here too - i.e. a supported string ('canvas', 'webgl', etc) or a constructor that inherits from `DrawerBase`.

iangilman reviewed (approved):
I'm impressed that one can add in a canvas drawer like this so cleanly! Pretty cool.

Adding `setDrawer` also seems like a good addition to the API.

I think this is a good approach; we want existing apps to work out of the box as much as possible, but having the warning also helps them update their code to work better.

I think we're good here except for the merge conflicts :)

pearcetm commented:
I've made a couple of last changes:
1) I renamed the new API `setDrawer` to `requestDrawer` since it doesn't always create and set the viewer's drawer, it just tries to create one, and only sets it conditionally now (See #2)
2) Consolidated all drawer creation code into `viewer.requestDrawer` so the logic is always the same in terms of options, etc. An optional argument specifies whether the new drawer should be used as the viewer's main drawer or not.
3) Added a new event `drawer-error` (raised on a `viewer`) that drawer implementations can fire to indicate problems drawing a tiled image.
4) Fixed the merge conflict, which was due to the recent patch to add `placeholderFillStyle` to the `WebGLDrawer`.

I believe we should be good to go now.

iangilman commented on src/viewer.js:
```diff
@@ -929,19 +929,20 @@ $.extend( $.Viewer.prototype, $.EventSource.prototype, $.ControlDock.prototype,
      * @param { Object } [drawerOptions] Options for this drawer. If falsey, defaults to viewer.drawerOptions
      * for this viewer type. See {@link OpenSeadragon.Options}.
-     * @returns {Boolean} whether the drawer was created successfully
+     * @returns {Object | Boolean} The drawer that was created, or false if the requestd drawer is not supported
```
Typo: requestd.

pearcetm replied:
Fixed.

iangilman commented on src/viewer.js:
```diff
@@ -929,19 +929,20 @@ $.extend( $.Viewer.prototype, $.EventSource.prototype, $.ControlDock.prototype,
+     * @returns {Object | Boolean} The drawer that was created, or false if the requestd drawer is not supported
      */
-    setDrawer(drawerCandidate, redrawImmediately = true, drawerOptions = null){
+    requestDrawer(drawerCandidate, mainDrawer = true, redrawImmediately = true, drawerOptions = null){
```
With this many parameters, does it make sense to take in an `options` object with named properties?

pearcetm replied:
I don't feel strongly one way or the other. I switched to an `options` object in the most recent commit.

iangilman replied:
Cool. I don't really have a good rule for when we should or shouldn't follow that pattern. This looks good though.

iangilman reviewed (changes requested):
Looks good to me! Just a couple super minor points...

Aiosa commented:
Btw new PRs do not run (again) the CI pipeline.

iangilman commented:
> Btw new PRs do not run (again) the CI pipeline.

@Aiosa Thank you for spotting this! I shall investigate...

iangilman reviewed (approved):
Thank you for fixing this up!
